### ISSUE TITLE
Reconcile the wrapper-only design principle with current torchtune recipe patches

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/cruijff_kit#465]
## Context

In conversation around #449 (PR #464), @msalganik articulated a design principle for cruijff_kit:

> All fine-tuning code is written and tested by torchtune. cruijff_kit's role is to build wrappers around that, not to take ownership of training-loop or training-adjacent code.

This is a strong correctness boundary: torchtune has tested its dataloader, distributed training, gradient accumulation, etc. If cruijff_kit patches the recipe, we now own the verification of the patched paths.

## Current state

The principle is the *target* but not the *current* state. There are **10 marked `# !--- cruijff_kit patch ---!` blocks per recipe**, in each of the three custom recipes:

- `lora_finetune_single_device_nightly.py`
- `lora_finetune_single_device_stable.py`
- `lora_finetune_distributed_stable.py`

= 30 patch blocks total. Examples include `stash_adapter_files` (movable — pure file shuffling after `save_checkpoint`), `epochs_to_save` filtering (inside the training loop), config reads, validation hooks. Some are movable to a post-job step or wrapper layer; some genuinely need to run inside the training loop and read training state.

With torchtune frozen (security patches only), these forks aren't a transitional state — they're our permanent vendored copy.

## Why this matters now

Issue #449 wanted to expose `num_workers` and `persistent_workers` as scaffold knobs to fix GPU underutilization. The DataLoader is constructed inside `_setup_data` with no hook. Under the principle, that change isn't shippable; without it, the issue can't fully close. PR #464 ships only the principle-clean piece (`cpus_per_task`); the rest is upstream-blocked-and-deferred to this conversation.

This pattern is going to recur. Every future feature that needs to reach into recipe code will hit the same fork. So we should decide deliberately, once:

## Three options for how to apply the principle going forward

1. **Strict.** Don't add new patches. Open follow-ups to extract every existing patch that's movable (probably ~half). Document the irreducible remainder as "we accept this divergence because no wrapper-layer mechanism exists." Performance ceilings that require recipe access (like #449's `num_workers`) become accepted limitations.

2. **Aspirational + case-by-case exceptions.** Each new patch needs an explicit "is this movable?" check. If movable → move it. If not, and the value is high → add it with a marked block, and the marked block is the audit trail. Existing patches stay unless someone has bandwidth to extract them.

3. **Status quo.** Don't add new patches *or* extract existing ones. Defer perf work that needs recipe access indefinitely.

## What I'd find useful from this discussion

- Which option (1/2/3) you want, and what triggers an exception under (2) if that's the choice.
- Whether the existing `stash_adapter_files` patch is (a) regret, (b) grandfathered, (c) within the principle as you read it. The answer changes the bar for similar future work.
- Whether to file a separate tracking issue for "extract `stash_adapter_files` from recipes" as a smoke-test for the strict path.
- Whether the principle should be written into `CLAUDE.md` so future Claudes (and humans) don't have to rediscover it case-by-case.

## Out of scope here

- Specific implementation of `num_workers` (covered by deferred work in #449).
- Changes to torchtune itself (frozen).
- General fork-management tooling.

cc @niznik-dev

**Comment by niznik-dev** [comment]
Two thoughts after just shipping #99, which is directly relevant — that PR added a fresh `# !--- cruijff_kit patch ---!` block to each of the three recipes (post-save `rewrite_adapter_config_base_path` call, plus a new `if save_adapter_weights_only else stash_adapter_files` branch).

**On `stash_adapter_files`:** I'd read it as location regret, not function regret. Both stash and the new rewrite are pure filesystem manipulation that runs after `save_checkpoint` returns and reads no training-loop state. If a post-save hook layer existed in the wrapper, both would live there cleanly. The functions are principle-compatible; their placement isn't.

**On which option:** I lean **(2)**, but with a more concrete trigger than just "is this movable?": *if a patch reads no training-loop state and runs strictly before-or-after a torchtune boundary call (`load_checkpoint`, `save_checkpoint`, `train`), it must move to the wrapper. Otherwise the marked block is the audit trail and the divergence is accepted.* That converts the principle into a coding rule that's easy to apply at review time.

Under that rule, the #99 patches are extractable to a post-save hook on the wrapper side. Happy to scope a follow-up PR if/when this lands.

—MxC

**Comment by msalganik** [comment]
@niznik-dev I don't totally understand what you wrote above. Stepping back from this issue, I'd like to be clear about what promise we are making to our users. Perhaps we should talk about this at the next software meeting.

**Comment by niznik-dev** [comment]
# Recipe Patches Audit — 2026-05-26

Audit of every `# !--- cruijff_kit patch ---!` block in the three custom
recipes (`src/tools/torchtune/custom_recipes/`), framed for the question:
**how many of the features we added to torchtune are still necessary?** For
policy framing (inside-`train()` vs. outside), see "Recipe Patching Policy
(#465)" in `CLAUDE.md`.

39 patch blocks total (13 per recipe) resolve into **six features**, all
common to all three recipes. Two additional patches don't appear in all
three — see *Recipe-specific patches* below.

## Features

| F# | Feature | stable | nightly | distributed | Still necessary? |
|---|---|---|---|---|---|
| F1 | **Sparse checkpoint schedule (`epochs_to_save`)** — init validation + post-epoch save schedule. | 18–25, 188–209, 885–910 | 18–25, 188–209, 962–987 | 17–24, 197–218, 1007–1026 | **Yes.** Wired through `setup_finetune.py` (argparse + validation), referenced in every experiment. The #465 success case. Keep. |
| F2 | **Adapter-save policy (#99, #495)** — post-checkpoint helpers `rewrite_adapter_config_base_path` / `stash_adapter_files`. | shares F1's 18–25 + 885–910 | shares F1's 18–25 + 962–987 | shares F1's 17–24 + 1007–1026 | **Yes.** Every fine-tune → eval pipeline depends on it. F2 rides on F1's surface area (no standalone blocks). |
| F3 | **Custom-metrics framework** — opt-in `calculate_custom_metrics(logits, labels, …)` called from `_loss_step`, wrapped in `try/except` that auto-disables on first exception. | 56–65, 213–217, 719–740, 845–850 | 56–65, 219–223, 749–770, **[916–919 unwrapped]** | 59–68, 222–226, 872–895, 966–971 | **Yes, but worth Matt's eye.** Framework is "opt-in" *only if the user removes the default file*. `src/utils/finetune_custom_metrics.py` ships with the repo, so the path runs every step on every fine-tune. No callers in blueprints or tests — only the recipes use it. Worth confirming the default metrics are still wanted; if obsolete, either delete the default file (true opt-in) or remove the framework. |
| F4 | **DataLoader perf knobs (`num_workers`, `persistent_workers`)** — #514, #515. | 378–381, 593–596, 641–644 | 384–387, 399–402†, 621–624, 669–672 | 382–385, 654–657, 701–704 | **Yes.** Just landed in v0.3.2 (#514, #515). †Nightly's 399–402 is val-dataloader plumbing — see *Recipe-specific* below. |
| F5 | **Quieter progress bars (`tqdm_miniters`)** — #269. | 167–169, 777–779 | 167–169, 848–850 | 175–177, 838–844 | **Yes, but evaluate.** Plumbed into all three default YAMLs (`1B_defaults.yaml`, `8B_defaults.yaml`, `finetune_template.yaml`). Behavior is purely visual. The `train()`-side patch is the one item in this audit that the policy flags as a high-risk-zone patch with low-risk content — either accept as a documented exception, or replace with torchtune's `TQDM_MININTERVAL` env var (which is what the patch originally displaced for ergonomic reasons). |
| F6 | **Recipe-scoped logger** — imports `cruijff_kit.utils.logger.setup_logger`, binds module-level `logger`. | 48–54 | 48–54 | 51–57 | **No — dead code in all three recipes.** `logger` is bound and never referenced; all actual logging routes through torchtune's `log = utils.get_logger("DEBUG")`. **History:** the logger came in via PR #161 (Oct 2025, project-wide `print()`→logging migration) and was *ported* into all three recipes as a standardization pass in Nov 2025. Its only actual in-recipe consumer was Dan Rigobon's embeddings-extraction work in nightly (#54 / #77, July 2025), which used `logger.info` to dump hidden tensors each batch. Stable and distributed got the import for symmetry but never had embeddings code, so they've been dead-code carriers from the start. The nightly embeddings code was removed in #516 as part of the #465 patching-policy cleanup (broken + incomplete + unused — see #517 for the post-mortem). If #517 ever lands, its redesigned spec writes tensors via `torch.save` rather than `logger.info`, so the binding won't return. Clean removal. |

## Recipe-specific patches

Patches that aren't in all three recipes:

| What | stable | nightly | distributed | Notes |
|---|---|---|---|---|
| F4 extension — validation dataloader plumbing | n/a | 399–402 | n/a | Nightly-only because only nightly has a val dataloader (`run_val_every_n_steps` is upstream torchtune-nightly, not a cruijff_kit patch). When `_distributed_nightly` lands (#474), this will need to be ported. |
| F3 marker drift — `train()` log-dict merge | wrapped (845–850) | **unwrapped (916–919)** | wrapped (966–971) | Same behavior in all three; only nightly is missing the `# !--- cruijff_kit patch ---!` wrapper. Two-line audit-trail fix. |

## Structural differences (not cruijff_kit patches)

For completeness — these explain layout differences across the three recipes
without being our drift:

- **`_loss_step` helper:** present in both single-device recipes, absent in
  distributed (inlined into `train()`). Drives the F3 location difference in
  distributed.
- **`validate()` method + `run_val_every_n_steps`:** upstream torchtune-nightly
  feature, present only in `_single_device_nightly`. Not patched by cruijff_kit.
- **No `_distributed_nightly` recipe yet:** tracked under #474.

## Recommended next actions

1. **Remove F6** from all three recipes. Three identical dead-code blocks; no behavior change.
2. **Fix the F3 marker drift** in nightly (wrap lines 916–919). Pure audit-trail fix.
3. **Verify F3 default metrics** — read `src/utils/finetune_custom_metrics.py` and decide: (a) keep as-is, (b) make it truly opt-in by deleting the default file, or (c) retire the framework entirely if no one's looking at the output.
4. **Decide F5** — accept `tqdm_miniters` as a documented in-`train()` exception under #465 (purely visual, no model-state read), or revert to `TQDM_MININTERVAL` env var.
5. **Confirm test coverage for F1 + F3 in-`train()` guards** under #465 — unit test for `validate_epochs_to_save`'s bad-input cases; integration test for the custom-metrics try/except.
6. **#474 prep:** when `_distributed_nightly` lands, it must carry F1–F5 plus the val-dataloader patch (F6 hopefully retired by then).

—MxC

**Comment by niznik-dev** [comment]
@msalganik I hope the above helps to break down what we talked about last Friday re: the features that we've added and whether or not we should keep them

**Comment by niznik-dev** [comment]
Following up on the "what promise do we make to our users" question — narrowing from the four candidate framings, **P2** and **P3** look like the strongest fit for what cruijff_kit already does, and they aren't mutually exclusive:

- **P2 — A safer, more scriptable torchtune for social-science research.** Patches are the value-add: init-time validation (F1 above), sparse checkpointing, a scriptable hook for custom metrics (F3), perf knobs (F4). The training *math* is still torchtune's; we add safety, scriptability, and an audit trail around it.
- **P3 — Reproducibility for our experiments.** Same config + same seed → same numbers. Patches don't have to be invisible; they have to be deterministic and logged.

The question I'd like your read on: do we want to **back the promise empirically** — i.e., add a test that runs the patched recipes vs. upstream torchtune (same config, same seed, same input) and asserts they agree to within a tolerance?

That converts the promise from a stated principle into a falsifiable check:

- For **P2**: a passing test means our patches don't move training math.
- For **P3**: the test itself *is* the reproducibility check, with a known-good (seed → loss-curve) pair pinned.
- It also directly answers the F4 open question in the audit: `num_workers` ordering effects, if real, would surface here. If the tolerance has to widen for F4 specifically, that's a real divergence we'd document or fix.

If this direction sounds right, I'd scope something small:
- One tiny config (1B + ~10 steps + small dataset) so cost is bounded.
- Bit-equality where possible (`num_workers=0`, deterministic algorithms on); ULP-bounded tolerance otherwise — measured before pinning a number.
- Cadence: nightly GPU smoke (`tests/integration/gpu/`) is a natural home.

Happy to file a follow-up issue with the scoping if P2 + P3 + a seeded-equivalence test feels like the right anchor. Or if you'd rather pick a different promise (P1 strict / P4 hands-off) we should redirect — flag it and we'll regroup.

—MxC

**Comment by niznik-dev** [comment]
For completeness on the alternatives I mentioned but didn't define above:

(**P1 — "same training math as torchtune":** wrapper-only; no in-`train()` patches. F3 and the in-loop side of F4 would have to come out.)

(**P4 — "a scientific workflow, not a training framework":** patches tolerated where unavoidable but treated as implementation details, not user-facing surface. Policy becomes about minimizing maintenance burden, not correctness claims.)

—MxC

**Comment by niznik-dev** [comment]
After talking to @msalganik , we now are thinking about strongly adhering to this wrapper principle.

(F1) This could be removed now that we save adapter weights and don't need the merged weights
(F2) This can be a separate post-torchtune step
(F3) This could be removed - we're not really using it
(F4) We will remove these - minimal gains anyway
(F5) We can try removing these and see if Claude still crashes
(F6) We think this is a relic of work Dan was doing and so it can be removed

So a new pull request will address these and test how it goes

**Comment by msalganik** [comment]
@niznik-dev given the complexity of this issue, do you want to separate it some to make it easier to review/test?  I'm happy to do a review with you on Fri afternoon.  Since some of this is close to core training, I want have extra care.

**Comment by niznik-dev** [comment]
Per-rip issues filed under milestone [**Wrapper-only principle (#465)**](https://github.com/niznik-dev/cruijff_kit/milestone/11):

- α F2 (#524) — wrapper layer (`rewrite_adapter_config_base_path`, `stash_adapter_files` → `post_finetune.py`) — **land first**
- β F1 (#525) — `epochs_to_save` / `save_last_epoch_only` / `validate_epochs_to_save` removed
- γ F3 (#526) — custom-metrics framework + `src/utils/finetune_custom_metrics.py` deleted
- δ F5 (#527) — `tqdm_miniters` removed (the "see how it goes" empirical item)
- ε F4 (#528) — `num_workers` / `persistent_workers` removed
- ζ F6 (#529) — dead `setup_logger` import block removed

Reference implementation: closed PR #523 (branch `465-strict-wrapper-principle` preserved). This issue stays open until all six land — at which point the recipes will be upstream-identical inside `train()` and the wrapper-only policy is in force.

—MxC

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Following up on your "what promise are we making to our users" question from earlier in this thread — here's a draft principle for `CLAUDE.md` that captures where we landed. Floating it for your read before I edit anything:

> 6. **Wrapper-only** - cruijff_kit integrates with external tools (torchtune, inspect-ai) by wrapping their public surfaces, not modifying their internals. The promise to users: training math stays upstream-faithful — same config and seed produce the same numbers torchtune would.

It would slot in as the sixth principle alongside Scientific / Modular / Practical / Privacy-respecting / Self-improving. The name matches the milestone we already agreed on; the elaboration is meant to be the direct answer to your question.

Two knobs worth your read:

- **Name**: "Wrapper-only" keeps continuity with the milestone. An alternative is "Upstream-faithful," which leads with the promise rather than the mechanism. I lean Wrapper-only for recognition value, but happy either way.
- **Scope**: drafted to name torchtune + inspect-ai explicitly. We could keep it generic ("external tools") so future additions inherit automatically — I think naming them is clearer right now but flag it for your call.

If the wording feels right, I'll open a small PR to add it to `CLAUDE.md`.

—MxC

**Comment by niznik-dev** [comment]
Principle phrasing decided on in person with @msalganik 

6. Wrapper-only - cruijff_kit integrates with external tools (torchtune, inspect-ai) by wrapping them, not modifying their internals

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Closing as complete: all six wrapper-only rips (α–ζ) are merged to main — the three custom recipes are back to upstream-identical inside `train()`, adapter housekeeping moved to the `post_finetune` wrapper, and the sixth **Wrapper-only** principle is live in `CLAUDE.md`. Nothing left under this umbrella.

—MxC
